hsmd/hsmd.c

```diff
@@ -1525,6 +1527,107 @@ static struct io_plan *handle_sign_invoice(struct io_conn *conn,
 			 take(towire_hsm_sign_invoice_reply(NULL, &rsig)));
 }
 
+/* stolen from gossipd/routing.c to parse wireaddr in node announcement msg*/
+static struct wireaddr *read_addresses(const tal_t *ctx, const u8 *ser)
+{
+	const u8 *cursor = ser;
+	size_t len = tal_count(ser);
+	struct wireaddr *wireaddrs = tal_arr(ctx, struct wireaddr, 0);
+
+	while (cursor && len) {
+		struct wireaddr wireaddr;
+
+		/* BOLT #7:
+		 *
+		 * The receiving node:
+		 *...
+		 *   - SHOULD ignore the first `address descriptor` that does
+		 *     NOT match the types defined above.
+		 */
+		if (!fromwire_wireaddr(&cursor, &len, &wireaddr)) {
+			if (!cursor)
+				/* Parsing address failed */
+				return tal_free(wireaddrs);
+			/* Unknown type, stop there. */
+			status_trace("read_addresses: unknown address type %u",
+				     cursor[0]);
+			break;
+		}
+
+		tal_arr_expand(&wireaddrs, wireaddr);
+	}
+	return wireaddrs;
+}
+
+static char *check_node_announcement_content(struct client *c, const u8 *ann)
+{
+	size_t offset = 66;
+	size_t len = tal_count(ann);
+	secp256k1_ecdsa_signature signature;
+	u32 timestamp;
+	u8 rgb_color[3];
+	u8 alias[32];
+	u8 *features, *addresses;
+	struct wireaddr *wireaddrs;
+	struct node_id node_id;
+	char *err = NULL;
+
+	/* check length of announcement */
+	if (len < offset) {
+		err = tal_fmt(tmpctx,
+			      "Node announcement too short");
+		goto done;
+	}
+
+	if (!fromwire_node_announcement(tmpctx, ann,
+					&signature, &features, &timestamp,
+					&node_id, rgb_color, alias,
+					&addresses)) {
+		err = tal_fmt(tmpctx,
+			      "Malformed node_announcement %s",
+			      tal_hex(tmpctx, ann));
+		goto done;
```

Comment:
why not just return the err instead of jumping?   e.g.:

```
    return tal_fmt(...)
```

Reply:
Yes! It's more direct